crypto/bn/bn_gcd.c

```diff
@@ -611,18 +612,28 @@ int BN_gcd(BIGNUM *r, const BIGNUM *in_a, const BIGNUM *in_b, BN_CTX *ctx)
         goto err;
 
     /* find shared powers of two, i.e. "shifts" >= 1 */
+    pow2_flag = 1;
+    pow2_shifts = 0;
+    pow2_numbits = 0;
     for (i = 0; i < r->dmax && i < g->dmax; i++) {
-        mask = ~(r->d[i] | g->d[i]);
-        for (j = 0; j < BN_BITS2; j++) {
-            bit &= mask;
-            shifts += bit;
-            mask >>= 1;
-        }
+        pow2_numbits_temp = r->d[i] | g->d[i];
+        pow2_condition_mask = ~(constant_time_is_zero_bn(pow2_flag)) & ~(constant_time_is_zero_bn(pow2_numbits_temp));
+        pow2_flag &= ~pow2_condition_mask;
```

Comment:
I prefer keep code clear, there is two conditions:
* flag != 0
* mask != 0

Compiler is clever enough to optimize this.

Reply:
I am not concerned about compiler.  Just trying to understand
code with triple negation is a bit hard for humans.
but how about
```C
 pow2_condition_mask = constant_time_is_zero_bn(pow2_numbits_temp);
 pow2_flag &= pow2_condition_mask;
```
wouldn't that also work, because pow2_flag is initialized with 1 and can only have one falling edge?

Reply:
> I am not concerned about compiler. Just trying to understand code with triple negation is a bit hard for humans.

Human should understand that `pow2_condition_mask` has `true` (`0xF...F` in this case) if condition to update met.

Updating flag is less understandable, but it's cost of consttime calculation.


> ```c
>  pow2_condition_mask = constant_time_is_zero_bn(pow2_numbits_temp);
>  pow2_flag &= pow2_condition_mask;
> ```
> 
> wouldn't that also work, because pow2_flag is initialized with 1 and can only have one falling edge?

This seems broke subsequent uses of `pow2_condition_mask`: it will be in *condition met* state more than once.

Reply:
> Hmm, now I see a way how to optimize the code even a bit more:
> 
> ```c
>     for (i = 0; i < r->dmax && i < g->dmax; i++) {
>         pow2_numbits_temp = r->d[i] | g->d[i];
>         pow2_condition_mask = constant_time_is_zero_bn(pow2_flag);
>         pow2_flag &= constant_time_is_zero_bn(pow2_numbits_temp);
>         pow2_shifts += pow2_flag;
>         pow2_numbits = constant_time_select_bn(pow2_condition_mask,
>                                                pow2_numbits, pow2_numbits_temp);
>     }
> ```
> 
> how about that?

Okay, this seems good.



> > I prefer keep code clear, there is two conditions: flag != 0 && mask != 0, not flag == 0 || mask == 0
> 
> Were those conditions in previous versions of this patch? When I look at the current version of the patch I don't really see them.

This expression was here since 3rd commit (when i removed `if(...) continue`), initially with `!= 0`, than `~constant_time_is_zero_bn`